web/src/components/users/RootAuthMethodsPage.tsx

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) [2024] SUSE LLC
+ *
+ * All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+import React, { useRef, useState } from "react";
+import {
+  Button,
+  FileUpload,
+  Flex,
+  Form,
+  FormGroup,
+  FormHelperText,
+  HelperText,
+  HelperTextItem,
+} from "@patternfly/react-core";
+import { useNavigate } from "react-router-dom";
+import { Center } from "~/components/layout";
+import { Page, PasswordInput } from "~/components/core";
+import { useRootUserMutation } from "~/queries/users";
+import { RootUserChanges } from "~/types/users";
+import { ROOT as PATHS } from "~/routes/paths";
+import { isEmpty } from "~/utils";
+import { _ } from "~/i18n";
+import textStyles from "@patternfly/react-styles/css/utilities/Text/text";
+import sizingStyles from "@patternfly/react-styles/css/utilities/Sizing/sizing";
+
+function RootAuthMethodsPage() {
+  const passwordRef = useRef();
+  const navigate = useNavigate();
+  const setRootUser = useRootUserMutation();
+  const [password, setPassword] = useState("");
+  const [sshKey, setSSHKey] = useState("");
+  const [isUploading, setIsUploading] = useState(false);
+
+  const startUploading = () => setIsUploading(true);
+  const stopUploading = () => setIsUploading(false);
+  const clearKey = () => setSSHKey("");
+
+  const isFormValid = !isEmpty(password) || !isEmpty(sshKey);
+  const uploadFile = () => document.getElementById("sshKey-browse-button").click();
+
+  const accept = async (e: React.SyntheticEvent) => {
+    e.preventDefault();
+
+    const data: Partial<RootUserChanges> = {};
+
+    if (!isEmpty(password)) {
+      data.password = password;
+      data.encryptedPassword = false;
+    }
+
+    if (!isEmpty(sshKey)) {
+      data.sshkey = sshKey;
+    }
+
+    if (isEmpty(data)) return;
+
+    await setRootUser.mutateAsync(data);
+
+    navigate(PATHS.root);
+  };
+
+  // TRANSLATORS: %s will be replaced by a link with the text "upload".
+  const [sshKeyStartHelperText, sshKeyEndHelperText] = _(
+    "Write, paste, drop, or %s a SSH public key file in the above textarea.",
+  ).split("%s");
```

Comment:
will this work also for right to left locales?

Reply:
Great question! Honestly, I don't know. But for now, I’m just using the interpolation we've documented at https://agama-project.github.io/docs/devel/i18n#formatting-texts.

That said, I really appreciate you raising concerns about RTL languages. While I did try to approach things with RTL in mind from the start, I’m afraid we're still far of having a fully RTL-compatible interface. There are so many parts involved, and even the CSS we're using is key.

Don't get me wrong. I’m not suggesting we stop thinking about RTL just because we don’t have everything in place yet. However, I'd keep things as they are for now and take notes on how to improve these translations to make them RTL-compatible, and tackle them all at once when we can.

Thanks so much for your input!

Reply:
!olleH
RTL expert here :-)

This `split("%s")` technique is fine with respect to bidirectional languages: there are just letters, no numbers or punctuation. Refer to https://github.com/mvidner/bidi-test for examples of failures related to that.

If done incorrectly (`"Write, paste, drop, or" + link + "a SSH public key file in the above textarea."`) it would be a problem because of _grammar_, not text direction:
In Japanese, the translation will come out like this:
> "In the above area, a SSH key write, paste, drop or %s".

Our implementation will handle that fine.

Reply:
Thanks a lot for the explanation @mvidner!